[clang-format] Add Automatic and ExceptShortType options for AlwaysBreakAfterReturnType.

rmarker opened the pull request:
The RTBS_None option in Clang-format avoids breaking after a short return type.
However, there was an issue with the behaviour in that it wouldn't take the leading indentation of the line into account.
This meant that the behaviour wasn't applying when intended.

In order to address this situation without breaking the existing formatting, RTBS_None has been deprecated.
In its place are two new options for AlwaysBreakAfterReturnType.
The option RTBS_Automatic will break after the return type based on PenaltyReturnTypeOnItsOwnLine.
The option RTBS_ExceptShortType will take the leading indentation into account and prevent breaking after short return types.

This allows the inconsistent behaviour of RTBS_None to be avoided and users to decide whether they want to allow breaking after short return types or not.

Resolves #78010

rmarker committed:
[clang-format] Add ShortReturnTypeColumn option.

rmarker commented on clang/include/clang/Format/Format.h:
```diff
@@ -3928,6 +3928,13 @@ struct FormatStyle {
+  /// regarded as short. Note that this isn't the length of the type itself,
+  /// but the column where it finishes. I.e. it includes indentation, etc.
+  /// \version 18
+  unsigned ShortReturnTypeLength;
```
Yeah, that's much nicer.

rmarker commented on clang/include/clang/Format/Format.h:
```diff
@@ -3928,6 +3928,13 @@ struct FormatStyle {
   /// \version 13
   unsigned ShortNamespaceLines;
 
+  /// When AlwaysBreakAfterReturnType is None, line breaks are prevented after
```
Done.

rmarker commented:
Merged main and fixed conflicts.

rmarker commented:
@mydeveloperday, do you have an opinion on this pull request?
It addresses #78010 with minimal changes, or are you interested in exploring other possibilities?

HazardyKnusperkeks commented:
Just ping me in a few days, if there is no other reaction and we merge it.

owenca commented:
> @mydeveloperday, do you have an opinion on this pull request? It addresses #78010 with minimal changes, or are you interested in exploring other possibilities?

It's weird to add such a top-level option to fix a bug and keep the current behavior. See https://github.com/llvm/llvm-project/issues/78010#issuecomment-1903300176.

rmarker commented:
Merged main and fixed conflicts.
Also updated version badge to 19 with the recent branching for 18.
Keeping things up to date whilst discussion about the underlying issue continues on #78010.

rmarker committed:
Take indentation into account for short return types.

rmarker committed:
Add AllowShortType option for AlwaysBreakAfterReturnType.

Replaced the new ShortReturnTypeColumn option with this alternative.

rmarker commented:
Reworked this PR with a new approach, as discussed in #78010.

`ShortReturnTypeColumn` has been replaced with a new enum option `RTBS_AllowShortType` for `AlwaysBreakAfterReturnType `.
Also, the short return type behaviour has been updated to ignore the leading indentation on the line.

rmarker committed:
Address review feedback.

rmarker committed:
Read the contents with encoding utf-8 when generating the format style.

On Windows it adding spurious characters to the output.

rmarker commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -587,7 +589,9 @@ bool ContinuationIndenter::mustBreak(const LineState &State) {
-       Style.AlwaysBreakAfterReturnType != FormatStyle::RTBS_None) &&
+       (Style.AlwaysBreakAfterReturnType != FormatStyle::RTBS_None &&
+        Style.AlwaysBreakAfterReturnType !=
+            FormatStyle::RTBS_AllowShortType)) &&
```
Nice.

rmarker commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9872,9 +9872,30 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
+               "  int g();\n"
+               "  long\n"
+               "  "
+               "fooooooooooooooooooooooooooooooooooooooooooooooo::"
```
Using `ColumnLimit` is a good idea.

rmarker commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9891,13 +9912,20 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "  g();\n"
+               "  long\n"
+               "  "
+               "fooooooooooooooooooooooooooooooooooooooooooooooo::"
```
Done.

rmarker commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9906,12 +9934,19 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "  int g();\n"
+               "  long\n"
+               "  "
+               "fooooooooooooooooooooooooooooooooooooooooooooooo::"
```
Done.

rmarker commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9920,13 +9955,20 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "  int g();\n"
+               "  long\n"
+               "  "
+               "fooooooooooooooooooooooooooooooooooooooooooooooo::"
```
Done.

rmarker commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9938,12 +9980,19 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "  int g();\n"
+               "  long\n"
+               "  "
+               "fooooooooooooooooooooooooooooooooooooooooooooooo::"
```
Done.

rmarker commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -326,11 +326,13 @@ bool ContinuationIndenter::canBreak(const LineState &State) {
+    // Don't break after very short return types (e.g. "void") as that is often
+    // unexpected.
+    if (Current.is(TT_FunctionDeclarationName) &&
+        State.Column - State.FirstIndent < 6) {
```
Yep.

rmarker commented on clang/docs/ClangFormatStyleOptions.rst:
```diff
@@ -146,7 +146,7 @@ the configuration (without a prefix: ``Auto``).
 .. _BasedOnStyle:
 
-**BasedOnStyle** (``String``) :ref:`¶ <BasedOnStyle>`
+**BasedOnStyle** (``String``) :ref:`Ã‚Â¶ <BasedOnStyle>`
```
Yes it was generated on Windows.
My guess is the platform is causing it to default to the wrong encoding.
I had a look at the script and discovered that forcing the encoding to utf-8 when reading the contents stopped it inserting the spurious characters.

rmarker commented:
Addressed feedback.

owenca commented:
This would break polly:
```
$ ninja polly-check-format
[27/105] Checking format of /Users/Owe...ude/polly/CodeGen/BlockGenerators.h...
FAILED: tools/polly/polly-check-format1 /Users/Owen/llvm-project/build/tools/polly/polly-check-format1 
cd /Users/Owen/llvm-project/build/tools/polly && /Users/Owen/llvm-project/build/bin/clang-format -sort-includes -style=llvm /Users/Owen/llvm-project/polly/include/polly/CodeGen/BlockGenerators.h | diff -u /Users/Owen/llvm-project/polly/include/polly/CodeGen/BlockGenerators.h -
--- /Users/Owen/llvm-project/polly/include/polly/CodeGen/BlockGenerators.h	2024-01-05 20:58:16
+++ -	2024-01-26 23:25:21
@@ -794,9 +794,9 @@
   /// @param BBMap A mapping from old values to their new values in this block.
   /// @param LTS   A mapping from loops virtual canonical induction variable to
   /// their new values.
-  void
-  generateScalarStores(ScopStmt &Stmt, LoopToScevMapT &LTS, ValueMapT &BBMAp,
-                       __isl_keep isl_id_to_ast_expr *NewAccesses) override;
+  void generateScalarStores(
+      ScopStmt &Stmt, LoopToScevMapT &LTS, ValueMapT &BBMAp,
+      __isl_keep isl_id_to_ast_expr *NewAccesses) override;
 
   /// Copy a single PHI instruction.
   ///
[36/105] Checking format of /Users/Owe...ject/polly/include/polly/ScopInfo.h...
FAILED: tools/polly/polly-check-format35 /Users/Owen/llvm-project/build/tools/polly/polly-check-format35 
cd /Users/Owen/llvm-project/build/tools/polly && /Users/Owen/llvm-project/build/bin/clang-format -sort-includes -style=llvm /Users/Owen/llvm-project/polly/include/polly/ScopInfo.h | diff -u /Users/Owen/llvm-project/polly/include/polly/ScopInfo.h -
--- /Users/Owen/llvm-project/polly/include/polly/ScopInfo.h	2024-01-22 03:28:44
+++ -	2024-01-26 23:25:21
@@ -1987,8 +1987,8 @@
   void addAccessData(MemoryAccess *Access);
 
   /// Add new invariant access equivalence class
-  void
-  addInvariantEquivClass(const InvariantEquivClassTy &InvariantEquivClass) {
+  void addInvariantEquivClass(
+      const InvariantEquivClassTy &InvariantEquivClass) {
     InvariantEquivClasses.emplace_back(InvariantEquivClass);
   }
 
ninja: build stopped: subcommand failed.
```

owenca commented:
OTOH, if we change `LLVMStyle.AlwaysBreakAfterReturnType` to `RTBS_AllowShortType`, we get the following:
```
$ ninja polly-check-format
[24/105] Checking format of /Users/Owe...olly/include/polly/DependenceInfo.h...
FAILED: tools/polly/polly-check-format16 /Users/Owen/llvm-project/build/tools/polly/polly-check-format16 
cd /Users/Owen/llvm-project/build/tools/polly && /Users/Owen/llvm-project/build/bin/clang-format -sort-includes -style=llvm /Users/Owen/llvm-project/polly/include/polly/DependenceInfo.h | diff -u /Users/Owen/llvm-project/polly/include/polly/DependenceInfo.h -
--- /Users/Owen/llvm-project/polly/include/polly/DependenceInfo.h	2024-01-05 20:24:43
+++ -	2024-01-27 02:13:13
@@ -332,8 +332,8 @@
 void initializeDependenceInfoPass(llvm::PassRegistry &);
 void initializeDependenceInfoPrinterLegacyPassPass(llvm::PassRegistry &);
 void initializeDependenceInfoWrapperPassPass(llvm::PassRegistry &);
-void initializeDependenceInfoPrinterLegacyFunctionPassPass(
-    llvm::PassRegistry &);
+void
+initializeDependenceInfoPrinterLegacyFunctionPassPass(llvm::PassRegistry &);
 } // namespace llvm
 
 #endif
[26/105] Checking format of /Users/Owe...polly/include/polly/LinkAllPasses.h...
FAILED: tools/polly/polly-check-format21 /Users/Owen/llvm-project/build/tools/polly/polly-check-format21 
cd /Users/Owen/llvm-project/build/tools/polly && /Users/Owen/llvm-project/build/bin/clang-format -sort-includes -style=llvm /Users/Owen/llvm-project/polly/include/polly/LinkAllPasses.h | diff -u /Users/Owen/llvm-project/polly/include/polly/LinkAllPasses.h -
--- /Users/Owen/llvm-project/polly/include/polly/LinkAllPasses.h	2024-01-05 20:58:16
+++ -	2024-01-27 02:13:13
@@ -138,8 +138,8 @@
 void initializeDependenceInfoPass(llvm::PassRegistry &);
 void initializeDependenceInfoPrinterLegacyPassPass(llvm::PassRegistry &);
 void initializeDependenceInfoWrapperPassPass(llvm::PassRegistry &);
-void initializeDependenceInfoPrinterLegacyFunctionPassPass(
-    llvm::PassRegistry &);
+void
+initializeDependenceInfoPrinterLegacyFunctionPassPass(llvm::PassRegistry &);
 void initializeIslAstInfoWrapperPassPass(llvm::PassRegistry &);
 void initializeIslAstInfoPrinterLegacyPassPass(llvm::PassRegistry &);
 void initializeCodeGenerationPass(llvm::PassRegistry &);
[33/105] Checking format of /Users/Owe...ect/polly/unittests/Isl/IslTest.cpp...
ninja: build stopped: subcommand failed.
```
which is an improvement IMO.

rmarker committed:
Switch LLVMStyle.AlwaysBreakAfterReturnType to RTBS_AllowShortType.

rmarker committed:
Update formatting in polly to pass polly-check-format.

rmarker commented:
Sounds good.

Updated formatting in polly.

owenca commented:
@llvm/pr-subscribers-clang-format this patch fixes a very old bug and will cause behavior changes whether the default is changed to the new `AllowShortType` or left at `None`. Which way should we go?

rmarker committed:
Update tests for change to llvm style.

rmarker commented:
Updated FormatTests to account for the change to the style.
(Forgetting them is what I get for coding late).

This should help when evaluating the scope of the change.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,23 @@ struct FormatStyle {
+    RTBS_AllowShortType,
     /// Always break after the return type.
     /// \code
     ///   class A {
```
Shouldn't it be added in this block also?

rmarker replied:
Yes, it should indeed be added.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,23 @@ struct FormatStyle {
     ///   };
     ///   int f();
     ///   int f() { return 1; }
+    ///   int foooooooooooooooooooooooooooooooooooooooooooooooo::
```
I'd go with short names, and just name them like `LongName` or similar and just break. This is documentation which should convey the settings, and must not break with LLVMs standard column limit.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,23 @@ struct FormatStyle {
+    ///       baaaaaaaaaaaaaaaaaaaaar();
     /// \endcode
     RTBS_None,
+    /// Break after return type automatically, while allowing a break after
+    /// short return types.
```
This should be rephrased, I know what the option means, but I couldn't figure it out of the description.

rmarker replied:
I made another attempt at being clearer.
What do you think?

HazardyKnusperkeks replied:
Better.

HazardyKnusperkeks commented on clang/lib/Format/ContinuationIndenter.cpp:
```diff
@@ -587,7 +590,7 @@ bool ContinuationIndenter::mustBreak(const LineState &State) {
       // Don't break before a C# function when no break after return type.
       (!Style.isCSharp() ||
-       Style.AlwaysBreakAfterReturnType != FormatStyle::RTBS_None) &&
+       Style.AlwaysBreakAfterReturnType > FormatStyle::RTBS_AllowShortType) &&
```
I know Owen proposed this. But I have to note, that I personally really don't like greater or lesser with enums.

owenca replied:
What I suggested is more consistent with the existing code in clang-format. Also, explicitly including/excluding every enum value is cumbersome and doesn't provide any real advantage, IMO.

HazardyKnusperkeks replied:
I know, just wanted to state my opinion. The cumbersome part could be extracted into a function. I just don't like treating enums as numbers in general. Sometimes there are exceptions, but in my opinion I would use this in clang-format only for `Standard`.

HazardyKnusperkeks commented:
> @mydeveloperday @HazardyKnusperkeks @rymiel this patch fixes a very old bug and will cause behavior changes whether the default is changed to the new `AllowShortType` or left at `None`. Which way should we go?

What would change if the default was kept at `None`? I don't see it.

owenca commented:
> What would change if the default was kept at `None`? I don't see it.

See https://github.com/llvm/llvm-project/pull/78011#issuecomment-1913052884.

owenca commented:
> @mydeveloperday @HazardyKnusperkeks @rymiel this patch fixes a very old bug and will cause behavior changes whether the default is changed to the new `AllowShortType` or left at `None`. Which way should we go?

Now I'm leaning toward keeping the existing (buggy) behavior of `None` and using the new `AllowShortType` to allow wrapping after short return types.

rmarker committed:
Revert "Read the contents with encoding utf-8 when generating the format style."

This reverts commit 285be69c1ebc2fd92659d45e86330ab19a416cc7.

rmarker commented on clang/docs/tools/dump_format_style.py:
```diff
@@ -474,7 +474,7 @@ class State:
 options_text = "\n\n".join(map(str, opts))
 
-with open(DOC_FILE) as f:
+with open(DOC_FILE, encoding="utf-8") as f:
```
Moved change to its own PR. #79805

rmarker committed:
Improve documentation.

rmarker commented:
> > @mydeveloperday @HazardyKnusperkeks @rymiel this patch fixes a very old bug and will cause behavior changes whether the default is changed to the new `AllowShortType` or left at `None`. Which way should we go?
> 
> Now I'm leaning toward keeping the existing (buggy) behavior of `None` and using the new `AllowShortType` to allow wrapping after short return types.

Yeah, it is a tricky trade-off.
The addition of `AllowShortType`, does alleviate the buggy behaviour and addresses the issue.

I can undo the fix for determining what types are short, if you think that is the way to go after the seeing the results.
I assume that this would also include changing the `LLVMStyle.AlwaysBreakAfterReturnType` back to `None`?

rmarker commented:
> > > @mydeveloperday @HazardyKnusperkeks @rymiel this patch fixes a very old bug and will cause behavior changes whether the default is changed to the new `AllowShortType` or left at `None`. Which way should we go?
> > 
> > 
> > Now I'm leaning toward keeping the existing (buggy) behavior of `None` and using the new `AllowShortType` to allow wrapping after short return types.
> 
> Buggy would be no break? That would be what I expect from `None`. :) So +1 from me.

`None` prevents breaks after short return types. The buggy behaviour is that it doesn't take leading indentation into account when determining if a return type is short or not. This means that if there is indentation, it wouldn't think a type is short that otherwise would be, and would thus allow a break after the return type.
Fixing the bug would cause return types in more situations to be identified as short and thus prevent breaks previously allowed (changing the previous behaviour).

 ```
 // Buggy None
void LongName::
      AnotherLongName();
class C {
   void
   LongName::AnotherLongName();
};

// Fixed None
void LongName::
      AnotherLongName();
class C {
   void LongName::
        AnotherLongName();
};
 ```

owenca commented:
Let's leave `None` alone and add `ExceptShortType`, e.g.:
```
--- a/clang/lib/Format/ContinuationIndenter.cpp
+++ b/clang/lib/Format/ContinuationIndenter.cpp
@@ -329,12 +329,18 @@ bool ContinuationIndenter::canBreak(const LineState &State) {
   // Don't break after very short return types (e.g. "void") as that is often
   // unexpected.
   if (Current.is(TT_FunctionDeclarationName) && State.Column < 6) {
     if (Style.AlwaysBreakAfterReturnType == FormatStyle::RTBS_None)
       return false;
   }
+  if (Current.is(TT_FunctionDeclarationName) &&
+      Style.AlwaysBreakAfterReturnType == FormatStyle::RTBS_ExceptShortType) {
+    assert(State.Column >= State.FirstIndent);
+    if (State.Column - State.FirstIndent < 6)
+      return false;
+  }
 
   // If binary operators are moved to the next line (including commas for some
   // styles of constructor initializers), that's always ok.
   if (!Current.isOneOf(TT_BinaryOperator, tok::comma) &&
       // Allow breaking opening brace of lambdas (when passed as function
       // arguments) to a new line when BeforeLambdaBody brace wrapping is
```

rmarker commented:
Right, I'll look into adding `ExceptShortType`.

rmarker committed:
Revert "Update tests for change to llvm style."

This reverts commit 64caa360077385479e7ad3791dc500674ab2f17a.

rmarker committed:
Revert "Update formatting in polly to pass polly-check-format."

This reverts commit 01fd15c864b01218d36f1acd86da8476cddf97c8.

rmarker committed:
Revert "Switch LLVMStyle.AlwaysBreakAfterReturnType to RTBS_AllowShortType."

This reverts commit 0fb332e1d8019ccf184b473e57a9762fdccaf5a0.

rmarker commented:
@owenca, thinking on it, do you think `NeverShortType` would be a better name?
I'm wondering whether `ExceptShortType` could imply that non-short return types are always broken after.

rmarker committed:
Add ExceptShortType option for AlwaysBreakAfterReturnType.

owenca commented:
> Let's leave `None` alone and add `ExceptShortType`

I meant to add `ExceptShortType` instead of `AllowShortType`. Sorry I didn't realize that adding `AllowShortType` and making it to mean "not `None`" might work, in which case we don't need to add `ExceptShortType`.

rmarker commented:
> > Let's leave `None` alone and add `ExceptShortType`
> 
> I meant to add `ExceptShortType` instead of `AllowShortType`. Sorry I didn't realize that adding `AllowShortType` and making it to mean "not `None`" might work, in which case we don't need to add `ExceptShortType`.

No problem.

Now that we've got both options, should we keep them?
Currently, `None` sometimes forbids breaking after short return types and sometimes allows it and just bases things off the penalty configuration. I feel like there will be some people who use `None` who would want the behaviour to be consistent one way and the other group wanting it consistent the opposite. With both, they can decide for themselves. Especially since the implementation for both is pretty simple and self-contained.

Though, if we only go with one, my vote would be for allowing breaking after short return types.
That is what we were using `None` for at work. I only discovered the issue, and the short return type concept at all, when investigating why it wasn't breaking in a few odd places we couldn't understand.

owenca commented:
> Now that we've got both options, should we keep them? Currently, `None` sometimes forbids breaking after short return types and sometimes allows it and just bases things off the penalty configuration. I feel like there will be some people who use `None` who would want the behaviour to be consistent one way and the other group wanting it consistent the opposite. With both, they can decide for themselves. Especially since the implementation for both is pretty simple and self-contained.

+1. In addition, we should deprecate `None` (in documentation only) and rename `AllowShortType` to `Automatic`, followed by `ExceptShortType` which means "`Automatic` except for short return types". Then IMO `ExceptShortType` is unambiguous. If you still think `NeverShortType` is clearer, I would change it to `NotShortType`.

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,39 @@ struct FormatStyle {
+    ///   int
+    ///   LongName::AnotherLongName();
+    /// \endcode
+    RTBS_AllowShortType,
```
```suggestion
    RTBS_Automatic,
```

rmarker replied:
Done

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,39 @@ struct FormatStyle {
     ///   };
     ///   int f();
     ///   int f() { return 1; }
+    ///   int LongName::
+    ///       AnotherLongName();
     /// \endcode
```
Change the comments (lines 917-927) to something like the following:
```
    /// This is **deprecated**. See ``Automatic`` below.
```

rmarker replied:
Done

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,39 @@ struct FormatStyle {
+    ///       AnotherLongName();
     /// \endcode
     RTBS_None,
+    /// Break after return type automatically.
+    /// This mode doesn't have the same inherent restriction on breaking after
+    /// short return types as ``RTBS_None`` and is solely based on
+    /// ``PenaltyReturnTypeOnItsOwnLine``.
```
For example:
```suggestion
    /// Break after return type based on ``PenaltyReturnTypeOnItsOwnLine``.
```

rmarker replied:
Done

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -922,8 +922,39 @@ struct FormatStyle {
+    ///   LongName::AnotherLongName();
+    /// \endcode
+    RTBS_AllowShortType,
+    /// Break after return type automatically.
+    /// ``PenaltyReturnTypeOnItsOwnLine`` is taken into account.
+    /// This mode will never break after short return types, unlike
+    /// ``RTBS_None`` which will only sometimes choose not to break after short
+    /// return types.
```
For example:
```suggestion
    /// Same as ``Automatic`` above, except that there is no break after short
    /// return types.
```

rmarker replied:
Done

owenca reviewed (commented):
There is no need to update the release notes for new option values.

rmarker committed:
Address review feedback.

Deprecate RTBS_None.
Rename RTBS_AllowShortType -> RTBS_Automatic.
Remove new option values from release notes.

rmarker commented:
> +1. In addition, we should deprecate `None` (in documentation only) and rename `AllowShortType` to `Automatic`, followed by `ExceptShortType` which means "`Automatic` except for short return types". Then IMO `ExceptShortType` is unambiguous. If you still think `NeverShortType` is clearer, I would change it to `NotShortType`.

Sounds like a plan.
Updated PR with changes.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9938,12 +9984,16 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "    return 1;\n"
                "  }\n"
                "  int g();\n"
+               "  long\n"
+               "  foooooooooooooooooooooooooooo::baaaaaaaaaaaaaaaaaaaar();\n"
```
Delete it because it's not a definition?

rmarker replied:
Deleted.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9938,12 +9984,16 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "}\n"
-               "int g();",
+               "int g();\n"
+               "int\n"
+               "foooooooooooooooooooooooooooo::baaaaaaaaaaaaaaaaaaaaar();",
```
Ditto.

rmarker replied:
Deleted.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9920,13 +9962,17 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
   verifyFormat("class C {\n"
                "  int f() { return 1; }\n"
                "  int g();\n"
+               "  long\n"
+               "  foooooooooooooooooooooooooooo::baaaaaaaaaaaaaaaaaaaar();\n"
```
Delete because it's not top-level?

rmarker replied:
Deleted.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9906,12 +9944,16 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
   verifyFormat("class B {\n"
                "  int f() { return 1; }\n"
                "  int g();\n"
+               "  long\n"
+               "  foooooooooooooooooooooooooooo::baaaaaaaaaaaaaaaaaaaar();\n"
```
Delete it because it's not a definition?

rmarker replied:
Yeah, not sure if this example is completely relevant in the tests for these subsequent options.
Though to me, it seems that each of the options has a test with the same block of code. Since there are other lines which aren't definitions in the same test.
Wondering if it is worth keeping it so that the symmetry in this area is preserved?
Though, I don't have strong feelings either way if you think it isn't needed.

owenca replied:
It's a declaration and supposed to have no break after the type. That's why I suggested that we delete it (and others).

rmarker replied:
Deleted.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -9906,12 +9944,16 @@ TEST_F(FormatTest, ReturnTypeBreakingStyle) {
                "}\n"
-               "int g();",
+               "int g();\n"
+               "int\n"
+               "foooooooooooooooooooooooooooo::baaaaaaaaaaaaaaaaaaaaar();",
```
Ditto.

owenca reviewed (commented):
Looks good otherwise.

rmarker committed:
Fix typo.

rmarker commented on clang/include/clang/Format/Format.h:
```diff
@@ -914,16 +914,31 @@ struct FormatStyle {
     /// \endcode
-    RTBS_None,
+    RTBS_Automatic,
+    /// Same as ``Automatic`` above, expect that there is no break after short
```
Oof, spellcheck has failed me ;)

rmarker committed:
Remove some superfluous code from tests.

rmarker commented:
Thanks, @owenca.
Don't know if you want to wait for any others to look it over.
I had a look at the buildkite check failure. Not sure if it is related to the changes in the PR?
In any case, I'll need someone to merge for me when the time is right.

owenca commented:
I don't think the buildkite failure is related to this PR, but I like to wait for a couple of days before merging in case others have any comments.

rmarker commented:
Nice.
Thanks, @owenca, and @HazardyKnusperkeks.
